frame/assets-freezer/src/lib.rs

```diff
@@ -0,0 +1,361 @@
+		// Figure out the most we can transfer from the balance on hold. This is basically the same
+		// as the balance on hold, but also ensures that the account actually has enough funds to
+		// be reduced and that no freezes have been placed on the asset/account.
+		let amount = T::Store::get(&(asset, who.clone())).reserved;
+		use fungibles::InspectWithoutFreezer;
+		let backing = T::Assets::reducible_balance(asset, who, true);
+		amount.min(backing)
+	}
+}
+
+impl<T: Config> fungibles::MutateHold<<T as SystemConfig>::AccountId> for Pallet<T> where
+	T::Assets: fungibles::Transfer<T::AccountId> + fungibles::InspectWithoutFreezer<T::AccountId>
+{
+	fn hold(asset: AssetIdOf<T>, who: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+		if !Self::can_hold(asset, who, amount) {
+			Err(TokenError::NoFunds)?
+		}
+		T::Store::mutate(
+			&(asset, who.clone()),
+			|extra| extra.reserved = extra.reserved.saturating_add(amount),
+		)?;
+
+		Self::deposit_event(Event::Held(asset, who.clone(), amount));
+		Ok(())
+	}
+
+	fn release(asset: AssetIdOf<T>, who: &T::AccountId, amount: BalanceOf<T>, best_effort: bool)
+		-> Result<BalanceOf<T>, DispatchError>
+	{
+		T::Store::try_mutate_exists(
+			&(asset, who.clone()),
+			|maybe_extra| if let Some(ref mut extra) = maybe_extra {
+				let old = extra.reserved;
+				extra.reserved = extra.reserved.saturating_sub(amount);
+				let actual = old - extra.reserved;
+				ensure!(best_effort || actual == amount, TokenError::NoFunds);
+
+				Self::deposit_event(Event::Released(asset, who.clone(), actual));
+
+				Ok(actual)
+			} else {
+				Err(TokenError::NoFunds)?
+			},
+		)
+	}
+
+	fn transfer_held(
+		asset: AssetIdOf<T>,
+		source: &T::AccountId,
+		dest: &T::AccountId,
+		amount: BalanceOf<T>,
+		on_hold: bool,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		// Can't create the account with just a chunk of held balance - there needs to already be
+		// the minimum deposit.
+		let min_balance = <Self as fungibles::Inspect<_>> ::minimum_balance(asset);
+		let dest_balance = <Self as fungibles::Inspect<_>>::balance(asset, dest);
+		ensure!(!on_hold || dest_balance >= min_balance, TokenError::CannotCreate);
+		Self::balance_on_hold(asset, dest).checked_add(&amount).ok_or(ArithmeticError::Overflow)?;
+
+		Self::decrease_on_hold_ensuring_backed(asset, source, amount)?;
+
+		// `death` is `KeepAlive` here since we're only transferring funds that were on hold, for
+		// which there must be an additional min_balance, it should be impossible for the transfer
+		// to cause the account to be deleted.
+		use fungibles::Transfer;
+		let result = Self::transfer(asset, source, dest, amount, WhenDust::KeepAlive);
+		if result.is_ok() {
+			if on_hold {
+				let r = Self::increase_on_hold(asset, dest, amount);
+				debug_assert!(r.is_ok(), "account exists and funds transferred in; qed");
+				r?
+			}
+		} else {
+			debug_assert!(false, "can_withdraw was successful; qed");
```

Comment:
This is an assumption that user must always call `can_withdraw` before calling `transfer_held`?

Why can't we directly call it ourselves above?

Reply:
It is called in `decrease_on_hold_ensuring_backed`